### PR TITLE
Extend to matrices and high-order tensors

### PR DESCRIPTION
**Pull request by jli05**
First clean-up:

* accessory image files moved into `assets/`
* demo notebooks moved into `demos/`
* use Python's built-in `unittest` package for unit tests, one dependency (on `pytest`) less

Next, I plan to extend the engine to matrix or higher-order tensors, and make the `backward()` more efficient so that it won't have to recompute the operator topology for each piece of new input data.

### CODE REVIEW

**Comment by dnparadice** [comment]
Why do you keep making PRs? The repo is fine the way it is. If you re-organize it, it will not match the video demos that go along with it.

**Comment by jli05** [comment]
This is an extension to matrices and higher-order tensors. The notebook demos still run.

To see how it is like, check out https://github.com/brief-ds/micrograd @dnparadice

**Comment by jli05** [comment]
@karpathy could you review? You can check it out at https://github.com/brief-ds/micrograd:

* extend the `Value` class and backward derivation to matrices and higher-order tensors
* add a minimal set of operators
* split the tests into `tests/test_engine.py` to test  deterministically (no dependency on torch) and `tests/test_vs_torch.py`

The `Value` class and `tensordot` function are exported at the package level:

```python
from micrograd import Value, tensordot
```

**Comment by jli05** [comment]
@karpathy this version works with matrices and higher-order tensors. Major benefits are:

* 24 kilobytes runtime code
* as portable as Python and NumPy
* loads 10x faster than PyTorch, 30x faster than TensorFlow

It does not break any of your previous demos. Could you consider merging, possibly up the version number in `setup.py` and release to PyPi?

**Comment by jli05** [comment]
@karpathy now the `demos/demo_scalar.ipynb` and `demos/demo_vector.ipynb` produce exactly the same 100 steps of training trace.

**Comment by jli05** [comment]
@karpathy you can check how it looks on https://github.com/brief-ds/micrograd

* 20kb core code at present (`__init__.py`, `engine.py` and `optim.py`)
* as portable as Python and NumPy
* loads 10x faster than PyTorch, 30x faster than TensorFlow
* the tensor-capable demo notebook `demos/demo_vector.ipynb` trains _instantly_, producing same intermediate results over 100 steps as your original demo notebook `demos/demo_scalar.ipynb`

**Comment by jli05** [comment]
@karpathy this version works with tensors.

As the core is just one 500-line Python file `micrograd/engine.py`, the learning curve is almost zero.

This blog post [TensorFlow, Apple's MLX and our micrograd](https://www.brief-ds.com/2025/09/25/tensorflow-mlx.html) explains in terms of install size and performance micrograd is in par with MLX, but micrograd is extra easy to learn, play with and profile.

Would you consider merging?

**Comment by jli05** [comment]
I made micrograd tensor-capable with in mind the idea to try simplifying the attention mechanism. Below is a proposal. Any thought? @karpathy 

https://www.brief-ds.com/2026/02/10/roadmap-att.html

Only micrograd's characteristics makes the study possible (simplicity, opening up forward and backward methods). Feel free to merge this PR. :)

**Comment by jli05** [comment]
@karpathy it seems the `max()` in `softmax()` should be mathematically derivable  as in microgpt

https://gist.github.com/karpathy/8627fe009c40f57531cb18360106ce95
